thetis/momentum_eq.py

```diff
@@ -297,31 +300,80 @@ def rhs(self, solution, eta, w=None, viscosity_v=None,
         if viscosity_h is not None:
-            f_visc = viscosity_h * (Dx(solution[0], 0) * Dx(self.test[0], 0) +
-                                    Dx(solution[1], 0) * Dx(self.test[1], 0) +
-                                    Dx(solution[0], 1) * Dx(self.test[0], 1) +
-                                    Dx(solution[1], 1) * Dx(self.test[1], 1))
+            def grad_h(a):
+                return as_matrix([[Dx(a[0], 0), Dx(a[0], 1), 0],
+                                  [Dx(a[1], 0), Dx(a[1], 1), 0],
+                                  [0, 0, 0]])
+
+            def outer_jump(v, n):
+                return outer(v('+'), n('+')) + outer(v('-'), n('-'))
+            visc_tensor = as_matrix([[viscosity_h, 0, 0],
+                                     [0, viscosity_h, 0],
+                                     [0, 0, 0]])
+
+            grad_uv = grad_h(solution)
+            grad_test = grad_h(self.test)
+            stress = visc_tensor*grad_uv
+            f += inner(grad_test, stress)*dx
+
             if self.horizontal_dg:
-                # interface term
-                mu_grad_sol = viscosity_h*nabla_grad(solution)
-                f += -(avg(mu_grad_sol[0, 0])*jump(self.test[0], self.normal[0]) +
-                       avg(mu_grad_sol[0, 1])*jump(self.test[1], self.normal[0]) +
-                       avg(mu_grad_sol[1, 0])*jump(self.test[0], self.normal[1]) +
-                       avg(mu_grad_sol[1, 1])*jump(self.test[1], self.normal[1]))*(dS_v+dS_h)
-                # TODO symmetric interior penalty term
-            f += f_visc * dx
+                assert self.h_elem_size is not None, 'h_elem_size must be defined'
+                # Interior Penalty method by
+                # Epshteyn (2007) doi:10.1016/j.cam.2006.08.029
+                # sigma = 3*k_max**2/k_min*p*(p+1)*cot(Theta)
+                # k_max/k_min  - max/min diffusivity
+                # p            - polynomial degree
+                # Theta        - min angle of triangles
+                # assuming k_max/k_min=2, Theta=pi/3
+                # sigma = 6.93 = 3.5*p*(p+1)
+                degree_h, degree_v = self.space.ufl_element().degree()
+                # TODO compute elemsize as CellVolume/FacetArea
+                # h = n.D.n where D = diag(h_h, h_h, h_v)
+                elemsize = (self.h_elem_size*(self.normal[0]**2 +
+                                              self.normal[1]**2) +
+                            self.v_elem_size*self.normal[2]**2)
+                sigma = 3.5*degree_h*(degree_h + 1)/elemsize
+                if degree_h == 0:
+                    raise NotImplementedError('horizontal visc not implemented for p0')
+                alpha = avg(sigma)
+                ds_interior = (dS_h + dS_v)
+                # oriented diffusive flux n.K.n = nu*(n_xy.n_xy)
+                # Pestiaux (2014)  DOI: 10.1002/fld.3900
+                ip_flux_jump = avg(dot(dot(self.normal, visc_tensor), self.normal))*jump(solution)
```

Comment:
I don't think this is the correct extension from a scalar diffusion scheme to vectorial viscosity. The term you implement gives;

    (v_+ . n_+ + v_- . n_-) * (u_+ . n_+ + u_- . n_-)

where v and u are the vectorial test and trial functions. This is the same as:

     (v_+ - v_-) . n_+n_+ . (u_+ - u_-)

This indicates that for a facet with normal direction in the x-direction (n=(1,0,0)) you only get a contribution for the u-component. Moreover for facet normals that are not Cartesian aligned the n_+n_+ tensor has off-diagonal components which indicates coupling between the components.

What you want however is a diffusion operator that acts independently on each velocity component (and is the same linear operator for each). This is obtained by:

  (v_+ n_+ + v_- n_-) : (u_+ n_+ + v_- n_-)

where : indicates the inner product that contracts the index of the v and u components, and the index of the normal components.

The way I think about it is that the jump in u is associated with an integration by parts of grad(u). grad(u) is a tensor however so the integration by parts gives a boundary term with the tensor outer product u n.

The same applies to the second line. I think that should be:

    f += inner(outer_jump(solution, self.normal), avg(grad_test)) * ds_interior

That's obviously ignoring the viscosity tensor so far. With viscosity I think the terms should be:

    f += alpha*inner(outer_jump(n, self.test), dot(avg(visc_tensor), outer_jump(self.normal, solution)))*ds_interior
    f += -inner(avg(dot(visc_tensor, nabla_grad(self.test))), outer_jump(self.normal, solution))*ds_interior
    f += -inner(outer_jump(self.normal, self.test), avg(dot(visc_tensor, nabla_grad(solution))))*ds_interior

Note that:
* I've swapped the normal and the solution in the outer_jump terms, and used nabla_grad instead of grad so that the first index is always the normal/derivative and the second is the velocity component. This is to ensure we contract with the correct index (the first) when contracting with the viscosity tensor
* If we contract with the viscosity tensor correctly, I don't think we need a grad_h (or nabla_grad_h) because the viscosity tensor will automatically project out the vertical direction.
* In all the terms we only contract the self.test index with the solution index, meaning that we indeed get the same operator that acts independently on each component.
* There's some subtleties with regards to the averaging of viscosity (does not matter for continuous viscosity) depending on the choice of the auxilary variable. I've here followed Pestiaux.

Please check my maths :-)